File: JAVA/118 - Pascal's Triangle.java
/*
 118. Pascal's Triangle (E)

 Approach 1 : 
 Time: O(numRows^2)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (numRows == 0) return res;
        
        // First layer always 1
        res.add(new ArrayList<Integer>());
        res.get(0).add(1);
        
        for (int i = 1; i < numRows; i++) {
            List<Integer> row = new ArrayList<Integer>();
            List<Integer> prevRow = res.get(i - 1);
            
            // First element
            row.add(1);
            
            for (int j = 1; j < i; j++) {
                row.add(prevRow.get(j - 1) + prevRow.get(j));
            }
            
            // Last element
            row.add(1);
            // Add row to list
            res.add(row);
        }
        return res;
    }
}

File: C++/118-Pascal's_Triangle.cpp
/*
 118. Pascal's Triangle (E)

 Approach 1 :
 Time: O(n)
 Space: O(n)

 Approach 2 : Brute force
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> res(numRows, vector<int>());
        for (int i = 0; i < numRows; i++) {
            res[i].resize(i + 1, 1);
            for (int j = 1; j < i; j++) {
                res[i][j] = res[i - 1][j - 1] + res[i - 1][j];
            }
        }
        return res;
    }
};

/* Approach 2 */
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        if (numRows == 0) return {};
        vector<vector<int>> res(numRows, vector<int>());

        // First layer
        res[0].push_back(1);

        for (int i = 1; i < numRows; i++) {
            // First element always 1
            res[i].push_back(1);

            for (int j = 1; j < i; j++) {
                res[i].push_back(res[i - 1][j - 1] + res[i - 1][j]);
            }

            // Last element always 1
            res[i].push_back(1);
        }
        return res;
    }
};